### ISSUE TITLE
Samsung Devices move AllDay Events around

### ISSUE THREAD

**Issue by albatrosify** [builttoroam/device_calendar#323]
I have gotten a few bug reports, that allDay Events are starting to appear on wrong days (often 1 day earlier than correct).
I did not change anything on my code, neither did this plugin change.
It looks like my app / the plugin is inserting the events correctly, as for 10 second around the elements appear correctly for example in google calendar. After this 10seconds they start to move a day earlier...

Anyone got any similar bug reports or has any idea on how to solve this? I am not finding any issues. iOS is fine, only android in Samsung devices (and i do not own any)

**Comment by thomassth** [comment]
Can you provide more details? Eg.
- what version are you using? Is it still affecting the latest prerelease?
- is it affecting local or synced calendars?
- Since it might be a vendor-specific problem, is there any difference if you view it afterwards via Samsung calendar or Google calendar?

**Comment by albatrosify** [comment]
I have to correct myself. It is no longer only Samsung Devices. The bug exists across devices.
I am using the latest null-safety dev release "^4.0.1-dev.13167"
So far I only know that this issue exists on synced calendars (namely Google Calendar)

I use an iPhone and do not have any problems with sync issues. Most of my users neither have problems even on android devices. 
I tried to reproduce the error on my development pixel 2 xl, as well as AVD Pixel 4 XL.
Everything was fine, UNTIL I changed the timezone in the settings of my google calendar (via web interface).
I switched it to another timezone, looked at the elements (via web interface) and then set it back to the correct time zone.

From this moment on, syncs from android to google calendar via this plugin move ONLY allday events.
Also it appears (when I look at the wrong elements afterwards from my iPhone) that the allday events spans from, for example: 17 Aug -> 16 Aug. This might explain, why the allDay Event shows a day earlier.

**Comment by thomassth** [comment]
Sounds like the problem is about retrieval of events from calendar.

Though I wonder if this is caused by some strange approach handled by Google or bug from the package itself.

We may need to write around the bug if it's the former.

**Comment by albatrosify** [comment]
Is anybody able to reproduce the problem?

**Comment by FCView** [comment]
Your issue may be related to the way allDay events are stored. They should be stored from midnight on the first day to midnight on the day following the last day.  For example, a 2-day allDay event on Jan 1-2 should be stored from Jan 1 00:00:00 to Jan 2 00:00:00. The are currently being stored as Jan 1 00:00:00 to Jan 1 00:00:00, and I believe they are using the local timezone, not UTC. See [this note about writing allDay events](https://developer.android.com/reference/android/provider/CalendarContract.Events#writing-to-events).

Check out my PR #324 and see if that fixes your issue.

Also, see [my other comment](https://github.com/builttoroam/device_calendar/issues/220#issuecomment-683474498), about problems with the way Android (or Android's CalendarContract) handles local calendars vs synced calendars. In my project, I ended up removing the ability to use local calendars for this reason.

**Comment by albatrosify** [comment]
With your PR I was able to somehow prevent this issue with the following code:
![image](https://user-images.githubusercontent.com/64252708/129798271-faad3c37-75a8-491a-894e-04fc0495a2ed.png)

Still it is really _scary_ to insert allDay Elements into the Calendar.

**Comment by SuperKrallan** [comment]
Yes, some things are not right with allDay events that span over more than one day. If you store an event that spans over 2 days, setting the allDay flag to true and the start time and end time like this (using the .toIso8601String() and print):
flutter: Start: 2022-01-14T00:00:00.000+0100
flutter: End: 2022-01-16T00:00:00.000+0100
...and save this event to the calendar, and then retrieve it again, both dates have lost their location information (so no +0100, but changed it's time with -1 h and added the "Z").
flutter: Start: 2022-01-13T23:00:00.000Z
flutter: End: 2022-01-14T22:59:59.000Z
Other **non allDay** events saved does not lose this information.
But can get the location information back by using the tTZDateTime.from() and add the location. Then you get...
flutter: Start: 2022-01-14T00:00:00.000+0100
flutter: End: 2022-01-14T23:59:59.000+0100
But we see here that the end time has changed from the saved 16th 00:00 to the 14th: 23:59. So instead of the exact 48h we had saved from the beginning we get back 23h 59 min 59 s...
So yes, if you are using the allDay events and save them to the calendar with start/end dates spanning over more than 24 h the end date will be changed, which in turn will be perceived as an error by the user. I think this is an error, I would like to be able to save multiple allDay events in this way, but I guess a workaround could be to make an allDay event, make sure the end date cant be set (or just hide it), and then repeat the event by setting the RecurrenceRule to the number of days you want.
